### DISCUSSION TITLE
Pimcore Studio Backend - OpenAPI Implementation

### DISCUSSION THREAD

**Discussion by dpfaffenbauer** [pimcore/pimcore#17156] [General]
Hey, I got a question about the OpenAPI Implementation. I am already thinking about how to integrate CoreShop into Studio and since CoreShop already uses sort of dynamic Serialization and Symfony Forms for Input Validations, it doesn't make lots of sense to create custom DTOs and hydrators to populate the data and create the OpenAPI Specification. Well, it would make sense, but it is loads of work :D.

I was researching and found this: https://symfony.com/bundles/NelmioApiDocBundle/current/index.html. With that I can re-use what's already there and also create the OpenAPI Specifications.

Now the question: Would it make sense for the BackendBundle to also use that Bundle? @fashxp Guess you the right person to ask that.

**Comment by alexz707** [comment]
Hello Dominik,

Thanks for your input. When we started with our new approach, we already had Nelmio in place, but we removed it for the following reasons:

- Nelmio is designed to work tightly with Doctrine. Since we are using our own ORM, we would only utilize a small percentage of this bundle but would still have another dependency.
- Nelmio is not designed for having multiple API docs, so if a custom bundle is already using it, there is a high chance it would get messed up.
- Nelmio also doesn't allow much customization. For example, security settings are only available globally, and you cannot change the logo.
- The documentation for Nelmio is also not the best.

That's why we have chosen the native OpenAPI approach.

Concerning your issue with adding your endpoints, I would suggest a PR to maybe extend either the class

https://github.com/pimcore/studio-backend-bundle/blob/f9089d1e6687f3e7cb471f5e68e30cbcc3d744c0/src/OpenApi/Service/OpenApiService.php#L35

and/or the class

https://github.com/pimcore/studio-backend-bundle/blob/f9089d1e6687f3e7cb471f5e68e30cbcc3d744c0/src/OpenApi/Controller/OpenApiController.php#L43

with an event where you can then hook into and add your endpoint config. This would also provide the possibility for others to modify the OpenAPI response to their needs.
What do you think about it?

Best,
Alex

**Reply by dpfaffenbauer** [reply]
@alexz707 Ok. Thats fine for me :)

Thanks
